presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/JoinPrefilter.java

```diff
@@ -0,0 +1,171 @@
+    @Override
+    public PlanOptimizerResult optimize(PlanNode plan, Session session, TypeProvider types, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        if (isEnabled(session)) {
+            Rewriter rewriter = new Rewriter(session, metadata, idAllocator, variableAllocator);
+            PlanNode rewritten = SimplePlanRewriter.rewriteWith(rewriter, plan, null);
+            return PlanOptimizerResult.optimizerResult(rewritten, rewriter.isPlanChanged());
+        }
+
+        return PlanOptimizerResult.optimizerResult(plan, false);
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final Session session;
+        private final Metadata metadata;
+        private final PlanNodeIdAllocator idAllocator;
+        private final VariableAllocator variableAllocator;
+        private boolean planChanged;
+
+        private Rewriter(Session session, Metadata metadata, PlanNodeIdAllocator idAllocator, VariableAllocator variableAllocator)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.metadata = requireNonNull(metadata, "functionAndTypeManager is null");
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.variableAllocator = requireNonNull(variableAllocator, "idAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<Void> context)
+        {
+            PlanNode left = node.getLeft();
+            PlanNode right = node.getRight();
+
+            PlanNode rewrittenLeft = rewriteWith(this, left);
+            PlanNode rewrittenRight = rewriteWith(this, right);
+            List<EquiJoinClause> equiJoinClause = node.getCriteria();
+
+            // We apply this for only left and inner join and the right side of the join is a simple scan and the join is on one key
+            if (equiJoinClause.size() == 1 &&
+                    (node.getType() == LEFT || node.getType() == INNER) &&
+                    isScanFilterProject(rewrittenLeft)) {
+                VariableReferenceExpression leftKey = equiJoinClause.stream().map(x -> x.getLeft()).findFirst().get();
+                VariableReferenceExpression rightKey = equiJoinClause.stream().map(x -> x.getRight()).findFirst().get();
+
+                // First create a SELECT DISTINCT leftKey FROM left
+                Map<VariableReferenceExpression, VariableReferenceExpression> leftVarMap = new HashMap();
+                PlanNode leftKeys = clonePlanNode(rewrittenLeft, session, metadata, idAllocator, ImmutableList.of(leftKey), leftVarMap);
+                PlanNode projectNode = projectExpressions(leftKeys, idAllocator, variableAllocator, ImmutableList.of(leftVarMap.get(leftKey)), ImmutableList.of());
+
+                // DISTINCT on the leftkey
+                PlanNode filteringSource = new AggregationNode(
```

Comment:
Isn't this redundant since the semijoin filters out duplicates?

Reply:
But the Bag datastructure ooms if you have giant right side

Reply:
In that case wouldn't it be better to include this additional distinct as a generic cost-based optimization to all semijoins with a large right side?